flatpak-import.py

```diff
@@ -0,0 +1,212 @@
+            OSTREE_STATIC_DELTA_SUPERBLOCK_FORMAT)
+    delta = GLib.Variant.new_from_bytes(
+            ostree_static_delta_superblock_format,
+            mapped_bytes,
+            False)
+
+
+    ### parse flatpak metadata
+
+    # use get_child_value instead of array index to avoid
+    # slowdown (constructing the whole array?)
+    checksum_variant = delta.get_child_value(3)
+    if not OSTree.validate_structureof_csum_v(checksum_variant):
+        # checksum format invalid
+        exit(1)
+
+    metadata_variant = delta.get_child_value(0)
+
+    commit = OSTree.checksum_from_bytes_v(checksum_variant)
+
+    ref = metadata_variant['ref']
+
+    if 'flatpak' in metadata_variant.keys():
+        flatpak = metadata_variant['flatpak']
+    else:
+        flatpak = ''
+
+    if 'origin' in metadata_variant.keys():
+        origin = metadata_variant['origin']
+    else:
+        origin = ''
+
+    if 'runtime-repo' in metadata_variant.keys():
+        runtime_repo = metadata_variant['runtime-repo']
+    else:
+        runtime_repo = ''
+
+    if 'metadata' in metadata_variant.keys():
+        app_metadata = metadata_variant['metadata']
+    else:
+        app_metadata = ''
+
+    if 'gpg-keys' in metadata_variant.keys():
+        gpg_keys = metadata_variant['gpg-keys']
+    else:
+        gpg_keys = ''
+
+    logging.debug("commit: " + commit)
+    logging.debug("flatpak: " + commit)
+    logging.debug("ref: " + ref)
+    logging.debug("origin: " + origin)
+    logging.debug("runtime_repo: " + runtime_repo)
+    logging.debug("app_metadata:\n---\n" + app_metadata + "---")
+    logging.debug("gpg_keys: " + gpg_keys)
+
+    logging.debug("metadata keys: " + str(metadata_variant.keys()))
+
+
+    # open repository
+
+    repo_file = Gio.File.new_for_path(args.repo)
+    repo = OSTree.Repo(path=repo_file)
+    if os.path.exists(os.path.join(args.repo, 'config')):
+        logging.info('Opening repo at ' + args.repo)
+        repo.open()
+    else:
+        logging.info('Creating archive-z2 repo at ' + args.repo)
+        os.makedirs(args.repo)
+        repo.create(OSTree.RepoMode.ARCHIVE_Z2)
+
+
+    # prepare transaction
+
+    repo.prepare_transaction(None)
+    repo.transaction_set_ref(None, ref, commit)
+
+
+    # execute the delta
+
+    flatpak_file = Gio.File.new_for_path(args.flatpak)
+    repo.static_delta_execute_offline(flatpak_file, False, None)
+
+
+    # verify gpg signature
+
+    if args.gpg_homedir:
+        gpg_homedir = Gio.File.new_for_path(args.gpg_homedir)
+    else:
+        gpg_homedir = None
+    if args.keyring:
+        trusted_keyring = Gio.File.new_for_path(args.keyring)
+    else:
+        trusted_keyring = None
+    gpg_verify_result = repo.verify_commit_ext(commit,
+                                               keyringdir=gpg_homedir,
+                                               extra_keyring=trusted_keyring,
+                                               cancellable=None)
```

Comment:
Poking around ostree, I think there's a way to get the detached commit metadata from the static delta and use that to verify the commit before executing the static delta. See `process_one_static_delta` in `src/libostree/ostree-repo-pull.c`. That would be nice so that an untrusted commit never actually gets commited to the object store.

But that would require re-implementing a few internal ostree tasks, so this is good for now.